### PR TITLE
[Feat] Add Scatterplot

### PR DESCRIPTION
**Pull request by justin-kleid**

### CODE REVIEW

**Comment by justin-kleid** [comment]
Work in progress :

- [x] Allow use to taskbar
- [x] Verify variable selection + scatterplot are done correct
- [x] Configure chatgpt to handle scatterpot
- [x] fix brushing UI to keep points highlighted.

File: geoda-ai/src/ai/openai-utils 2.ts
Comment (lixun910) [inline geoda-ai/src/ai/openai-utils 2.ts:1]: It looks like there are many duplicated files named as "xxxx 2.tsx".

File: geoda-ai/src/components/chatgpt/chatgpt-panel.tsx [hunk 1]
```diff
@@ -37,7 +37,8 @@ const ChatGPTPanel = () => {
   const tableName = useSelector((state: GeoDaState) => state.root.file?.rawFileData?.fileName);
 
   // get api key from state.root
-  const openAIKey = useSelector((state: GeoDaState) => state.root.uiState.openAIKey);
+  //const openAIKey = useSelector((state: GeoDaState) => state.root.uiState.openAIKey);
+  const openAIKey = process.env.NEXT_PUBLIC_OPENAI_API_KEY;
```
Comment (lixun910) [inline geoda-ai/src/components/chatgpt/chatgpt-panel.tsx]: You don't need to change this line: the process.env.NEXT_PUBLIC_OPENAI_API_KEY has been used when initialize the redux store, see [ui-reducer.ts](https://github.com/GeoDaCenter/reactgeoda/blob/f46904aeea65eddb483b2a81e0c31b8a875259fb/geoda-ai/src/reducers/ui-reducer.ts#L3)

File: geoda-ai/src/components/chatgpt/custom-scatter-message.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  const dispatch = useDispatch();
+  const [hide, setHide] = useState(false);
+  const {output} = props;
+
+  if (!output.data || !Array.isArray(output.data) || output.data.length === 0) {
+    console.error('Scatterplot data is unavailable or invalid.');
+    return null;
+  }
+
+  const scatterplotData = output.data[0] as ScatterplotOutput['data'][0]; // Assuming data is an array with a single ScatPlotDataProps object
```
Comment (lixun910) [inline geoda-ai/src/components/chatgpt/custom-scatter-message.tsx]: The result should be a single scatter plot, so I think there is no need to use Array here.

File: geoda-ai/src/components/plots/plot-management.tsx [hunk 3]
```diff
@@ -71,6 +80,24 @@ export const PlotManagementPanel = () => {
...
+            </div>
+          }
+          className="p-2"
+        >
+          {plots
+            .filter(plot => plot.type === 'scatter')
+            .toReversed()
+            .map(plot => {
+              if (isScatterPlot(plot)) {
+                return <Scatterplot key={plot.id} data={plot.data[0]} />;
```
Comment (lixun910) [inline geoda-ai/src/components/plots/plot-management.tsx]: Same here: no need to use plot.data[0], just plot.data or plot?

File: geoda-ai/src/components/plots/scatterplot-panel.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            variant="solid"
+            color="warning"
+            selectedKey={showPlotsManagement ? 'plot-management' : 'scatterplot-creation'}
+            onSelectionChange={onTabChange}
+          >
+            <Tab key="scatterplot-creation" title="Create Scatterplot">
+              <Card>
+                <CardBody>
+                  <div className="flex flex-col gap-4">
+                    <MultiVariableSelector setVariables={setSelectedVariables} />
```
Comment (lixun910) [inline geoda-ai/src/components/plots/scatterplot-panel.tsx]: Use two Single Variable selector for "Independent Variable X" and "Dependent) variable Y"

File: geoda-ai/src/utils/custom-functions.ts [hunk 5]
```diff
@@ -99,6 +100,17 @@ export type HistogramOutput = {
...
 
+export type ScatterplotOutput = {
+  type: 'scatter';
+  name: string;
+  result: {
+    variableX: string;
+    variableY: string;
+    points: Array<{x: number; y: number}>;
+  };
+  data: ScatPlotDataProps[];
```
Comment (lixun910) [inline geoda-ai/src/utils/custom-functions.ts]: Same: data: ScatPlotDataProps;

File: geoda-ai/src/utils/scatterplot-utils.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  const points = xData.map((x, index) => ({x, y: yData[index]}));
+
+  return {
+    variableX,
+    variableY,
+    points
+  };
+}
+
+// const ChartSettings = {
```
Comment (lixun910) [inline geoda-ai/src/utils/scatterplot-utils.ts]: Good idea to move the function of getting scatter plot props here in the scatterplot-utils.ts I prefer to put it here if there is no other issue 👍

**Review by lixun910 — APPROVED** [review]
Looks great and good job! You can land this pr after addressing the comments. Thanks!

File: geoda-ai/src/utils/scatterplot-utils.ts (hunk 6 above)
Comment (justin-kleid) [reply geoda-ai/src/utils/scatterplot-utils.ts]: Sounds good - I'll take a look over the comments and merge tonight!

**Pull request merged**